internal/auth/oidc/auth_method.go

```diff
@@ -0,0 +1,193 @@
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+	"github.com/hashicorp/go-kms-wrapping/structwrapping"
+	"google.golang.org/protobuf/proto"
+)
+
+// DefaultAuthMethodTableName defines the default table name for an AuthMethod
+const DefaultAuthMethodTableName = "auth_oidc_method"
+
+// AuthMethod contains an OIDC auth method configuration. It is owned
+// by a scope.  AuthMethods can have may assigned Accounts, AudClaims,
+// CallbackUrls, Certificates, SigningAlgs.  AuthMethods also have one State at
+// any given time which determines it's behavior for many its operations.
+type AuthMethod struct {
+	*store.AuthMethod
+	tableName string
+}
+
+// NewAuthMethod creates a new in memory AuthMethod assigned to scopeId.
+// WithMaxAge, WithName and WithDescription are the only valid options. All
+// other options are ignored.
+//
+// State equals the state of the OIDC auth method.  State is not a supported
+// parameter when creating new AuthMethod's since it must be Inactive for all
+// new AuthMethods.
+//
+// DiscoveryUrl equals a URL where the OIDC provider's configuration can be
+// retrieved. The OIDC Discovery Specification requires the URL end in
+// /.well-known/openid-configuration. However, Boundary will strip the URI and
+// only store the scheme and domain.
+//
+// ClientId equals an OAuth 2.0 Client Identifier valid at the Authorization
+// Server.
+//
+// ClientSecret equals the client's secret which will be encrypted when stored
+// in the database and an hmac representation will also be stored when every the
+// secret changes.  The secret is not returned via the API, the hmac is returned
+// so callers can determine if it's been updated.
+//
+// MaxAge equals the Maximum Authentication Age. Specifies the allowable elapsed
+// time in seconds since the last time the End-User was actively authenticated
+// by the OP. If the elapsed time is greater than this value, the OP MUST
+// attempt to actively re-authenticate the End-User. A value -1 basically
+// forces the IdP to re-authenticate the End-User.  Zero is not a valid value.
+//
+// See: https://openid.net/specs/openid-connect-core-1_0.html
+func NewAuthMethod(scopeId string, discoveryUrl *url.URL, clientId string, clientSecret ClientSecret, opt ...Option) (*AuthMethod, error) {
+	const op = "oidc.NewAuthMethod"
+
+	var u string
+	switch {
+	case discoveryUrl != nil:
+		u = discoveryUrl.String()
+	}
+
+	opts := getOpts(opt...)
+	a := &AuthMethod{
+		AuthMethod: &store.AuthMethod{
+			ScopeId:      scopeId,
+			Name:         opts.withName,
+			Description:  opts.withDescription,
+			State:        string(InactiveState),
+			DiscoveryUrl: u,
+			ClientId:     clientId,
+			ClientSecret: string(clientSecret),
+			MaxAge:       int32(opts.withMaxAge),
+		},
+	}
+	if err := a.validate(op); err != nil {
+		return nil, err // intentionally not wrapped.
+	}
+	if a.ClientSecretHmac != "" {
+		return nil, errors.New(errors.InvalidParameter, op, "client secret hmac should be empty")
+	}
```

Comment:
Is this necessary since `ClientSecretHmac` is never set in the above code?

Reply:
Depends... I know it should never be set.  If a future dev doesn't understand that and sets it, then there will be a defect.  If I add this here, then future dev (aka. Me in a few months) will surely hesitate before setting it above.